src/flow/src/utils.rs

```diff
@@ -0,0 +1,465 @@
+                    continue;
+                }
+                s.update_event_ts(now, &key)?;
+            }
+            // the first batch with key that's greater or equal to ts
+            let batch = if let Some((_, batch)) = self.spine.range_mut(ts..).next() {
+                batch
+            } else {
+                // if no batch with `batch key >= ts`, then create a new batch with key being `ts`
+                self.spine.entry(ts).or_default()
+            };
+
+            {
+                let key_updates = batch.entry(key).or_insert(smallvec![]);
+                key_updates.push((val, ts, diff));
+            }
+        }
+        Ok(())
+    }
+
+    /// advance time to `now` and consolidate all older(`now` included) updates to the first key
+    pub fn set_compaction(&mut self, now: Timestamp) -> Result<(), EvalError> {
+        let mut should_compact = self.spine.split_off(&(now + 1));
+        std::mem::swap(&mut should_compact, &mut self.spine);
+
+        // if a full arrangement is not needed, we can just discard everything before and including now
+        if !self.full_arrangement {
+            return Ok(());
+        }
+        // else we update them into current key value pairs
+        let mut compacted_batch: BTreeMap<Row, SmallVec<[DiffRow; 2]>> = Default::default();
+
+        for (_, batch) in should_compact {
+            for (key, updates) in batch {
+                if let Some(s) = &mut self.expire_state {
+                    if s.is_key_expired(now, &key)? {
+                        continue;
+                    }
+                    s.update_event_ts(now, &key)?;
+                }
+                // if diff cancel out each other, then remove the key
+                let mut old_row: Option<DiffRow> =
+                    compacted_batch.get(&key).and_then(|v| v.first()).cloned();
+
+                for new_row in updates {
+                    old_row = compact_diff_row(old_row, new_row);
+                }
+                if let Some(compacted_update) = old_row {
+                    compacted_batch.insert(key, smallvec![compacted_update]);
+                } else {
+                    compacted_batch.remove(&key);
+                }
+            }
+        }
+
+        // insert the compacted batch into spine with key being `now`
+        let mut first_entry = BTreeMap::from([(now, compacted_batch)]);
+        self.spine.append(&mut first_entry);
+        Ok(())
+    }
+
+    /// get the updates of the arrangement from the given range of time
+    pub fn get_updates_in_range<R: std::ops::RangeBounds<Timestamp>>(
+        &self,
+        range: R,
+    ) -> Vec<KeyValDiffRow> {
+        let mut result = vec![];
+        for (_ts, batch) in self.spine.range(range) {
+            for (key, updates) in batch.clone() {
+                for (val, ts, diff) in updates {
+                    result.push(((key.clone(), val), ts, diff));
+                }
+            }
+        }
```

Comment:
Can we eliminate some `clone`s here?

Reply:
> Can we eliminate some `clone`s here?

The second condition need to return a local variable as get's result(it's like a multiversion thing that get value at a given time) so I'm afraid I have to clone in here

Reply:
get it